### PR TITLE
[Service Connector] `az webapp/containerapp/spring connection create/update`: Support opt out action

### PR DESCRIPTION
**Pull request by wchigit**
**Related command**
az webapp connection create <target-type> --source-id <source-id> --target-id <target-id> --opt-out publicnetwork configinfo
az webapp connection update <target-type> --source-id <source-id> --target-id <target-id> --opt-out publicnetwork configinfo

**Description**
This PR enables an opt-out option during the creation and the updates of Service Connector. Users can choose to skip the configuration settings on source service and the public network configuration on target service by adding arguments after '--opt-out'.  This parameter is optional.

**Testing Guide**
Create a service linker for App Service and Storage Blob without configuring public network and app settings:
az webapp connection create storage-blob --connection {connection-name} --source-id {source-id} --target-id {target-id} --secret --client-type dotnet --opt-out publicnetwork configinfo

Create a service linker for App Service and Storage Blob without configuring firewall rules on storage:
az webapp connection create storage-blob --connection {connection-name} --source-id {source-id} --target-id {target-id} --secret --client-type dotnet --opt-out publicnetwork

Update a service linker with no configuration changes on source
az webapp connection update storage-blob --id {conneciton-id} --client-type dotnet --opt-out configinfo

**History Notes**
[Service Connector] `az webapp/containerapp/spring connection create/update`: Support opt out action for configurationInfo and publicNetwork

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
Service Connector

File: src/azure-cli/azure/cli/command_modules/serviceconnector/_params.py [hunk 1]
```diff
@@ -236,6 +236,16 @@ def add_confluent_kafka_argument(context):
                      help='Name of the connection', validator=validate_kafka_params)
 
 
+def add_opt_out_argument(context):
+    context.argument('opt_out_set', options_list=['--opt-out'],
```
Comment (houk-ms) [inline src/azure-cli/azure/cli/command_modules/serviceconnector/_params.py:240]: can we make the `arg_type` of the list be an enum? something like
```
arg_type=get_enum_type(option_out_options)
```

File: src/azure-cli/azure/cli/command_modules/serviceconnector/action.py [hunk 2]
```diff
@@ -213,3 +213,24 @@ def get_action(self, values, option_string):  # pylint: disable=no-self-use
...
+
+
+class AddOptOut(argparse.Action):
+    def __call__(self, parser, namespace, values, option_string=None):
+        action = self.get_action(values)
+        namespace.opt_out_set = action
+
+    def get_action(self, values):  # pylint: disable=no-self-use
+        opt_out_set = set()
+        opt_out_options = {'public-network', 'config'}
```
Comment (houk-ms) [inline src/azure-cli/azure/cli/command_modules/serviceconnector/action.py:225]: Can the `authentication` step be opted-out?

can private-endpoint and private-link be otped-out? how about just using `network` and `configuration`?

Comment (wchigit) [reply src/azure-cli/azure/cli/command_modules/serviceconnector/action.py:225]: for private-endpoint and private-link, our api does not have an opt-out option. I will look into authentication opt-out.

Comment (yungezz) [reply src/azure-cli/azure/cli/command_modules/serviceconnector/action.py:225]: `config` is so ambiguous, could it be `configinfo` to align with swagger to give user some hint?

Comment (yungezz) [reply src/azure-cli/azure/cli/command_modules/serviceconnector/action.py:225]: > Can the `authentication` step be opted-out?
> 
> can private-endpoint and private-link be otped-out? how about just using `network` and `configuration`?

vnetsolution is by default off.

Comment (wchigit) [reply src/azure-cli/azure/cli/command_modules/serviceconnector/action.py]: > Can the `authentication` step be opted-out?
> 
> can private-endpoint and private-link be otped-out? how about just using `network` and `configuration`?

I don't think authentication can be opted out in our creation workflow.

File: src/azure-cli/azure/cli/command_modules/serviceconnector/_resource_config.py [hunk 3]
```diff
@@ -1104,3 +1104,10 @@ class CLIENT_TYPE(Enum):
 SUPPORTED_CLIENT_TYPE[RESOURCE.ContainerApp] = SUPPORTED_CLIENT_TYPE[RESOURCE.WebApp]
 SUPPORTED_CLIENT_TYPE[RESOURCE.Local] = SUPPORTED_CLIENT_TYPE[RESOURCE.WebApp]
 SUPPORTED_CLIENT_TYPE[RESOURCE.FunctionApp] = SUPPORTED_CLIENT_TYPE[RESOURCE.WebApp]
+
+
+# The dict defines the options for opt-out
+class OPT_OUT_OPTION(Enum):
+    PUBLIC_NETWORK = 'public-network'
```
Comment (yungezz) [inline src/azure-cli/azure/cli/command_modules/serviceconnector/_resource_config.py:1111]: naming convention for argument name is x-x-.., but no need for argument value, how about to remove the hyphen for user convenince,

File: src/azure-cli/azure/cli/command_modules/serviceconnector/_resource_config.py [hunk 4]
```diff
@@ -1104,3 +1104,10 @@ class CLIENT_TYPE(Enum):
 SUPPORTED_CLIENT_TYPE[RESOURCE.ContainerApp] = SUPPORTED_CLIENT_TYPE[RESOURCE.WebApp]
 SUPPORTED_CLIENT_TYPE[RESOURCE.Local] = SUPPORTED_CLIENT_TYPE[RESOURCE.WebApp]
 SUPPORTED_CLIENT_TYPE[RESOURCE.FunctionApp] = SUPPORTED_CLIENT_TYPE[RESOURCE.WebApp]
+
+
+# The dict defines the options for opt-out
+class OPT_OUT_OPTION(Enum):
+    PUBLIC_NETWORK = 'publicnetwork'
+    CONFIGURATION_INFO = 'configinfo'
```
Comment (houk-ms) [inline src/azure-cli/azure/cli/command_modules/serviceconnector/_resource_config.py:1112]: how about just using `network` and `appsettings` ?

File: src/azure-cli/azure/cli/command_modules/serviceconnector/action.py (hunk 2 above)
Comment (houk-ms) [reply src/azure-cli/azure/cli/command_modules/serviceconnector/action.py]: then for network part, i think we should use `network`, as the opt-out option covers both public network and vnet

for config part, how about use `appsettings`? `configinfo` seems also ambiguous

File: src/azure-cli/azure/cli/command_modules/serviceconnector/_resource_config.py (hunk 4 above)
Comment (wchigit) [reply src/azure-cli/azure/cli/command_modules/serviceconnector/_resource_config.py:1112]: `network` sounds fine to me, though it sounds like private endpoint solution is involved. `firewall` is another option, what do you think?
`appsettings` might be a little exclusive to web app?

**Comment by wchigit** [comment]
> @kairu-ms Could you please review this PR? Thanks.

Kind reminder. @kairu-ms Thanks.

**Comment by jsntcy** [comment]
As Kai Ru will take long leave, @necusjz, @AllyW, could you please help review this PR?

**Review by AllyW — APPROVED** [review]

**Comment by AllyW** [comment]
> **Related command** az webapp connection create --source-id --target-id --opt-out publicnetwork configinfo az webapp connection update --source-id --target-id --opt-out publicnetwork configinfo
> 
> **Description** This PR enables an opt-out option during the creation and the updates of Service Connector. Users can choose to skip the configuration settings on source service and the public network configuration on target service by adding arguments after '--opt-out'. This parameter is optional.
> 
> **Testing Guide** Create a service linker for App Service and Storage Blob without configuring public network and app settings: az webapp connection create storage-blob --connection {connection-name} --source-id {source-id} --target-id {target-id} --secret --client-type dotnet --opt-out publicnetwork configinfo
> 
> Create a service linker for App Service and Storage Blob without configuring firewall rules on storage: az webapp connection create storage-blob --connection {connection-name} --source-id {source-id} --target-id {target-id} --secret --client-type dotnet --opt-out publicnetwork
> 
> Update a service linker with no configuration changes on source az webapp connection update storage-blob --id {conneciton-id} --client-type dotnet --opt-out configinfo
> 
> **History Notes**
> 
> [Component Name 1] BREAKING CHANGE: `az command a`: Make some customer-facing breaking change [Component Name 2] `az command b`: Add some customer-facing feature
> 
> This checklist is used to make sure that common guidelines for a pull request are followed.
> 
> * [x]  The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).
> * [x]  I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).
> * [x]  I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

@wchigit  please update the  **History Notes** part in pr description, for later release doc generation, according to this guideline https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests

**Comment by jsntcy** [comment]
@yanzhudd, need your review as an owner.

**Review by necusjz — APPROVED** [review]

**Comment by wchigit** [comment]
> @yanzhudd, need your review as an owner.

Gentle reminder. @yanzhudd

**Review by yanzhudd — APPROVED** [review]

**Comment by wchigit** [comment]
@necusjz @AllyW @yanzhudd Thanks for reviewing. Could you help merge this PR?

**Pull request merged**